Fixing up unrolling to Reductions to use reorder

kevinstephano opened the pull request:
Apply reordering of Reduction Schedule in the magic scheduler on unrolling dimensions.  This allows the unroll to be split last but put at the highest dimension of the tensor for easier application of computeAt()

kevinstephano committed:
Apply reordering of Reduction Schedule in the magic scheduler on unrolling dimensions.  This allows the unroll to be split last but put at the highest dimension of the tensor for easier application of computeAt()

kevinstephano committed:
Fix clang format.

tlemo commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -415,15 +415,18 @@ c10::optional<ReductionParams> scheduleReduction(
-      // Unroll a certain number of rFactored elements
-      // Split Grid dimension to get multiple reds per block
+      red_tv->split(1, kLoopUnrollSplit);
+      red_tv->reorder({{-1, -2}, {-2, -1}});
```
a comment would help here to explain the logic behind the reorder

csarofeen replied:
Can you increase the computeAt point because of the reorder?

tlemo commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -445,51 +453,52 @@ c10::optional<ReductionParams> scheduleReduction(
         red_tv->split(1, rparams.grid_dim_y_);
         red_tv->split(1, rparams.block_dim_y_);
+        red_tv->split(1, kLoopUnrollSplit);
+        red_tv->reorder({{-1, -4}, {-4, -1}});
```
same

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -437,6 +440,11 @@ c10::optional<ReductionParams> scheduleReduction(
       red_tv_rf->axis(-2)->parallelize(ParallelType::TIDx);
       red_tv_rf->axis(-1)->parallelize(ParallelType::Unroll);
 
+      red_tv->axis(0)->parallelize(ParallelType::BIDx);
```
Did you reparallelize this? Looks like it's duplicated

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -445,51 +453,52 @@ c10::optional<ReductionParams> scheduleReduction(
 
         auto red_tv_rf = red_tv->rFactor(
             {-5, -1}); // NOLINT(cppcoreguidelines-avoid-magic-numbers)
         red_tv_rf->computeAt(red_tv, 1);
```
wondering if you can also increase this computeAt point now

csarofeen reviewed (approved):
No major objections, some minor comments.

kevinstephano commented:
I added a few more things:

- Thorough commenting of dimension splitting/reordering.
- I added better computeAt()s for rFactored Tensors such that less binding is required.

kevinstephano committed:
Added better computeAt() statements for rFactored Tensors of reductions and removed unneeded bindings.  Added thorough comments.
